Adding workflow for CI main upstream trigger

dimpar opened the pull request:
This workflow is intended to be called from CI main upstream workflow (`workflow_dispatch`) and switch to `allthekeeps-on-ropsten` branch.

dimpar committed:
Adding workflow so CI upstream can trigger it

lukasz-zimnoch commented:
Can we name the workflow file to `allthekeeps-testnet.yaml` to keep it consistent with what we have in other repos?

lukasz-zimnoch commented on .github/workflows/allthekeeps-ropsten.yml:
```diff
@@ -0,0 +1,115 @@
+    runs-on: ubuntu-latest
+    outputs:
+      path-filter: ${{ steps.filter.outputs.path-filter }}
+    steps:
```
Don't we miss the `dorny/paths-filter@v2` step here?

michalinacienciala replied:
Actually, if we don't intend to limit the execution of workflow for PRs to PRs affecting particular paths, we can get rid of the whole `dapp-detect-changes` job.
The `...-detect-changes` job is needed only if we want to execute some other job in the workflow only for PRs affecting specific paths and make that other job required for a PR merge.

dimpar replied:
~Actually, I think we can get rid of this job. It's a standalone project and I think we can run the workflow on every change.~ Ha, I had same thoughts what @michalinacienciala suggested, she was faster with her comment :)

lukasz-zimnoch commented on .github/workflows/allthekeeps-ropsten.yml:
```diff
@@ -0,0 +1,115 @@
+      - name: Run install repositories
+        run: ./install-repositories.sh
+      
+      - name: Get upstream packages' versions
```
Unnecessary `'`

lukasz-zimnoch commented on .github/workflows/allthekeeps-ropsten.yml:
```diff
@@ -0,0 +1,115 @@
+          CONTRACT_OWNER_ETH_ACCOUNT_PRIVATE_KEY=${{secrets.CONTRACT_OWNER_ETH_ACCOUNT_PRIVATE_KEY}} \
+          ETH_RPC_URL=${{secrets.ETH_RPC_URL}} \
+          SUBGRAPH_DEPLOY_KEY=${{secrets.SUBGRAPH_DEPLOY_KEY}} \
+          SUBGRAPH_SLUG=${{secrets.SUBGRAPH_SLUG}} ./deploy-subgraph.sh
```
Can we move `./deploy-subgraph.sh` to a separate line for better visibility?

dimpar replied:
done

lukasz-zimnoch commented:
Is there any particular reason why do we want to keep all the changes on `allthekeeps-on-ropsten` branch? Can't we just merge everything to `master`?

dimpar commented:
> Is there any particular reason why do we want to keep all the changes on `allthekeeps-on-ropsten` branch? Can't we just merge everything to `master`?

I thought about it, but I've made some [small changes](https://github.com/miracle2k/allthekeeps/compare/master...keep-network:allthekeeps-on-ropsten?expand=1) in the prod code to hardcode and point to Ropsten. In the future we might expect some new changes in the atk upstream code and I wanted to avoid resolving conflicts on "master" branches when we pull updates to our forked version of atk. Basically, I wanted to keep it clean and completely separate our own changes from the forked master branch.

michalinacienciala commented on .github/workflows/allthekeeps-ropsten.yml:
```diff
@@ -0,0 +1,115 @@
+      
+      - uses: actions/setup-node@v2
+        with:
+          node-version: "12.x"
```
Could we use version `14.x`? We're systematically upgrading to `14.x` where it's possible.
Also, we could add built-in caching mechanism for yarn dependencies
```suggestion
          node-version: "14.x"
          cache: "yarn"
```

dimpar replied:
I've tried, but the current version of other dependencies specified in the `package.json` won't work on 14.x

dimpar replied:
Added cache

michalinacienciala commented on .github/workflows/allthekeeps-ropsten.yml:
```diff
@@ -0,0 +1,115 @@
+            keep-ecdsa-contracts-version = github.com/keep-network/keep-ecdsa/solidity#version
+            tbtc-contracts-version = github.com/keep-network/tbtc/solidity#version
+
+      - name: Deploy subgraph to thegraph.com
```
Is it safe to deploy the subgraph to `thegraph.com` when workflow gets triggered by `push` or `pull_request`?
If we first run the `allthekeeps-ropsten.yml` as a part of the CI flow (i.e. using `workflow_dispatch` event) we will deploy a subgraph and then install dapp and perform publish to GCP bucket.
And if after some time we crate a PR in the `allthekeeps` project, this will result in deploying the subgraph again (which may differ from the previous one) and we will not push anything new to the GCP registry.
Will that not break the AllTheKeeps? (I'm not fully grasping the role of the subgraph and how is it constructed, so I'm not sure)

dimpar replied:
If we trigger `./deploy-subgraph.sh` it will indeed deploy a new version of subgraph to the `thegraph.com`, however, the version of atk will still point to the previous version (cause the previous version of the 'build' folder was deployed to GCP). `Push` or `pull requests` will validate everything except the deployment to GCP, but it won't break the dApp which was deployed by `workflow_dispatch` event.

dimpar committed:
Improving readability

michalinacienciala commented on .github/workflows/allthekeeps-ropsten.yml:
```diff
@@ -0,0 +1,115 @@
+name: Allthekeeps / Testnet
```
```suggestion
name: AllTheKeeps / Testnet
```

dimpar replied:
Done

dimpar committed:
Allthekeeps -> AllTheKeeps

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+        default: 'main'
+
+jobs:
+  dapp-build-and-publish:
```
```suggestion
  build-and-publish:
```

dimpar replied:
Done

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+on:
+  push:
+    branches:
+     - allthekeeps-on-ropsten
```
We should use master (main).

dimpar replied:
pls see this comment: https://github.com/keep-network/allthekeeps/pull/1#issuecomment-900255001

nkuba replied:
Oh, now I see that  I had prepared a comment regarding https://github.com/keep-network/allthekeeps/pull/1#issuecomment-900255001 but it got lost.

> > Is there any particular reason why do we want to keep all the changes on `allthekeeps-on-ropsten` branch? Can't we just merge everything to `master`?
> 
> I thought about it, but I've made some [small changes](https://github.com/miracle2k/allthekeeps/compare/master...keep-network:allthekeeps-on-ropsten?expand=1) in the prod code to hardcode and point to Ropsten. In the future we might expect some new changes in the atk upstream code and I wanted to avoid resolving conflicts on "master" branches when we pull updates to our forked version of atk. Basically, I wanted to keep it clean and completely separate our own changes from the forked master branch.


We should keep everything in master branch. If it requires changes to support testnet along with the mainnet we should introduce them to make the master code more universal. We should consider contributing the changes to the original repository, so we don't face conflicts in the future.

nkuba replied:
This is not blocking this PR, but should be handled in a follow-up.

dimpar replied:
Ok, I'd rather not block this PR. I agree, that making it more universal would be better/cleaner, but I figured that if all we need is to make it work on Ropsten for now then it would be good enough for our purposes. The last commit to the original ATK was made 6months ago so it's not something that changes often. I see what I can do to make it more universal in the follow up PRs

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+
+jobs:
+  dapp-build-and-publish:
+    if: github.event_name != 'pull_request'
```
In the case of `pull_request` we should build but not publish, it looks like we don't execute the workflow at all now for `pull_request`.

dimpar replied:
We do the same thing for `tbtc dapp` & `keep dashboard`. You think running the flows "on push" won't be enough?

michalinacienciala replied:
Kuba's right here. For `tbtc dapp` & `keep dashboard` we have
```
    if: |
      github.event_name != 'pull_request'
        || needs.dapp-detect-changes.outputs.path-filter == 'true'
```
and the latter part of this condition is responsible for execution of job for `pull_request` (`path-filter` is set to `true` for workflows triggered by PRs affecting filtered paths).
In the `allthekeeps` workflow we want to execute the job for all pull requests, no matter which files are modified in them.

dimpar replied:
ok, we can run on pull request as well. https://github.com/keep-network/allthekeeps/pull/1/commits/a350667bf11ed6b1aafd499edf852528d93ba5ea

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          ref: allthekeeps-on-ropsten
```
We will be able to remove it when we decide to use master.

dimpar replied:
Pls see the comment above

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+name: AllTheKeeps / Testnet
```
Can we have one workflow for any network `.github/workflows/allthekeeps.yml`?
```suggestion
name: AllTheKeeps
```

dimpar replied:
This particular workflow handles specifically Ropsten network. All the scripts that are triggered and dApp is pointing to Ropsten. ATK on mainnet is in hands of community.

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+      - name: Load environment variables
+        uses: keep-network/ci/actions/load-env-variables@v1
+        with:
+          environment: 'ropsten'
```
We could try to use `environment` workflow_dispatch input variable.

dimpar replied:
sorry, not sure what you mean. FWIW we have the same pattern across all other dApps in keep org.

michalinacienciala replied:
Kuba means using `environment: '${{ github.event.inputs.environment }}'`.
We don't use that for other dApps, but we do in other workflows, for example [here](https://github.com/keep-network/keep-core/blob/c92d1d627323a3bf98c2aa8397bd9ca696bfb453/.github/workflows/contracts.yml#L128).
It wasn't used in dApps because prior to uploading dApp files to GCS, we were using the `Load environment variables` action not only on `workflow_dispatch` but also on other events (so we couldn't read `environment` from the `workflow_dispatch` input, as it would be empty for workflows run due to push or PR). 
Right now, when switched to GCS, we could use the `environment: '${{ github.event.inputs.environment }}'` for all our dApps.

dimpar replied:
By other dApps I meant that we have similar hardcoded network name in [keep-dashboard](https://github.com/keep-network/keep-core/blob/c92d1d627323a3bf98c2aa8397bd9ca696bfb453/.github/workflows/dashboard-testnet.yml#L59) and [tbtc-dapp](https://github.com/keep-network/keep-core/blob/c92d1d627323a3bf98c2aa8397bd9ca696bfb453/.github/workflows/dashboard-testnet.yml#L59). I'll change it to `environment: '${{ github.event.inputs.environment }}'`

dimpar replied:
763c068

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+            keep-ecdsa-contracts-version = github.com/keep-network/keep-ecdsa/solidity#version
+            tbtc-contracts-version = github.com/keep-network/tbtc/solidity#version
+
+      - name: Deploy subgraph to thegraph.com
+        if: github.event_name != 'workflow_dispatch'
```
What is a use case for this step? Is it intended to be executed on master merge? Do we expect it to use contract packages versions from `package.json`? I'm not sure about it. Don't we want to publish on `workflow_dispatch` only?

dimpar replied:
In this step we generate a new subgraph for Ropsten purposes and in order to verify that it was generated successfully we deploy it to The Graph. We do not deploy a new version of ATK to GCP in this step, only to The Graph.

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+          SUBGRAPH_SLUG=${{secrets.SUBGRAPH_SLUG}} \
+            ./deploy-subgraph.sh
+
+      - name: Deploy latest subgraph to thegraph.com
+        if: github.event_name == 'workflow_dispatch'
```
Can we move this step right below `Get upstream packages versions` to keep `if: github.event_name == 'workflow_dispatch'` together? It seems cleaner to the flow as now we are mixing the ifs:
```
      - name: Get upstream packages versions
        if: github.event_name == 'workflow_dispatch'
...
      - name: Deploy subgraph to thegraph.com
        if: github.event_name != 'workflow_dispatch'
...
      - name: Deploy latest subgraph to thegraph.com
        if: github.event_name == 'workflow_dispatch'
```

dimpar replied:
Done

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+        run: ./install-dapp.sh
+
+      - name: Deploy to GCP bucket
+        if: github.event_name == 'workflow_dispatch'
```
We deploy to GCP only on `workflow_dispatch`. Maybe we should do the same for `Deploy subgraph to thegraph.com`?

dimpar replied:
Practically, I think that in most cases `Deploy subgraph to thegraph.com` won't be triggered, cause atk is a forked repo and there were no plans (as far as I know) to work on the code. I've added `Deploy subgraph to thegraph.com` mostly to check that the dots (deploying to The Graph, building everything except publishing the dApp) are connecting when we run into a situation when we want to test some minor changes and I don't think it's necessary to publish on GCP when testing manually. In most cases `workflow_dispatch ` is an event that will trigger everything.

nkuba replied:
Please correct me if I'm wrong but with the current configuration:
https://github.com/keep-network/allthekeeps/blob/763c068eac54e1f2174a3ad23b90097925d0285c/.github/workflows/allthekeeps-testnet.yml#L3-L20
We will deploy a new subgraph on each merge and pull_request.

I vote for removing `pull_request` and `push` trigger and the following step:
```
      - name: Deploy subgraph to thegraph.com
        if: github.event_name != 'workflow_dispatch'
```
So that this workflow `allthekeeps-testnet.yaml` will handle only `workflow_dispatch` events, that will be used for a testnet version release. If we want to test the code we should create a separate workflow or make this one a generic network-agnostic workflow.

There is already a `frontend.yml` workflow that is used to test app building.

dimpar replied:
ok, done 3f59695623de142141a74bc77c7ebe98db1b4f79

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+        env:
+          GITHUB_TOKEN: ${{ secrets.CI_GITHUB_TOKEN }}
+        with:
+          module: "https://github.com/keep-network/allthekeeps"
```
This should match ci configuration file https://github.com/keep-network/ci/pull/15/files#diff-ffb6b648a5b9ecb316c6b166af7cfc006edc09675cf2be200ad799180c92446aR51
```suggestion
          module: "github.com/keep-network/allthekeeps"
```

dimpar replied:
Done

dimpar committed:
Renames, reordering flows and adding cache

dimpar committed:
Running a flow on pull request

dimpar committed:
Replacing hardcoded network name with input variable

nkuba commented on .github/workflows/allthekeeps-testnet.yml:
```diff
@@ -0,0 +1,103 @@
+        run: ./install-repositories.sh
+      
+      - name: Deploy subgraph to thegraph.com
+        if: github.event_name != 'workflow_dispatch'
```
We should remove this step as I mentioned in the other comment. We shouldn't deploy a new subgraph on any `pull_request`. If we want to deploy it on a `push` to a particular master we could have it in a separate workflow.

dimpar replied:
same, done in 3f59695

dimpar committed:
Leaving workflow only for workflow_dispatch events

For the testing purposes of the dApp on push & pull_requests we can move
the steps to a seperate workflow.
